hostinfo/hostinfo_linux.go

```diff
@@ -26,6 +26,48 @@ func init() {
 	if v := linuxDeviceModel(); v != "" {
 		SetDeviceModel(v)
 	}
+	if d := linuxDesktop(); d != "" {
+		SetDesktop(d)
+	}
+}
+
+// linuxDesktop attempts to detect the class of desktop being used by the
+// system, if any. Given that tailscaled is likely to be run as a system
+// service, the usual environment-variable checks are not possible. Instead,
+// we look for unix sockets typical of a desktop Linux setup.
+func linuxDesktop() string {
+	var (
+		seenWayland bool
+		seenX       bool
+		seenSway    bool
+		seenDbus    bool
+	)
+
+	err := lineread.File("/proc/net/unix", func(line []byte) error {
+		seenDbus = seenDbus || strings.Contains(string(line), " @/tmp/dbus-")
+		seenX = seenX || strings.Contains(string(line), ".X11-unix")
+		seenSway = seenSway || strings.Contains(string(line), "/sway-ipc.")
+		seenWayland = seenWayland || strings.Contains(string(line), "/wayland-1")
+		return nil
+	})
+	if err != nil {
+		return ""
+	}
+
+	// Ordered such to indicate the most specific desktop first.
+	if seenSway {
+		return "sway"
+	}
+	if seenWayland {
+		return "wayland"
+	}
+	if seenX {
+		return "X11"
+	}
+	if seenDbus {
+		return "user-dbus"
+	}
```

Comment:
I thought we were just going to do a bool, to answer the question of whether it's a Linux server or a Linux desktop.

How important is Sway vs Wayland vs X11? That's a pretty esoteric detail. Why not GNOME vs KDE vs XFCE at that point, for instance?

If we do a bool, it should be a tristate (opt.Bool), though, so we can distinguish old clients that didn't check.

Reply:
My original thinking was that X11 vs Wayland would be useful information, if we ever wanted to make a GUI or tray client on Linux. Regardless, I've changed this to be an `opt.Bool`.